src/server/session.ts

```diff
@@ -531,26 +531,35 @@ namespace ts.server {
             defaultDefinition :
             defaultProject.getLanguageService().getSourceMapper().tryGetSourcePosition(defaultDefinition!));
 
-        // Track which projects we have already searched so that we don't repeat searches.
-        // We store the project key, rather than the project, because that's what `loadAncestorProjectTree` wants.
-        // (For that same reason, we don't use `resultsMap` for this check.)
-        const searchedProjects = new Set<string>();
+        // The keys of resultsMap allow us to check which projects have already been searched, but we also
+        // maintain a set of strings because that's what `loadAncestorProjectTree` wants.
+        const searchedProjectKeys = new Set<string>();
 
         onCancellation:
         while (queue.length) {
             while (queue.length) {
                 if (cancellationToken.isCancellationRequested()) break onCancellation;
 
+                let skipCount = 0;
+                for (; skipCount < queue.length && resultsMap.has(queue[skipCount].project); skipCount++);
+
+                if (skipCount === queue.length) {
+                    queue.length = 0;
+                    break;
+                }
+
+                if (skipCount > 0) {
+                    queue.splice(0, skipCount);
+                }
+
+                // NB: we may still skip if it's a project reference redirect
                 const { project, location } = queue.shift()!;

```

Comment:
Would it make more sense long-term not to de-queue at all? Just iterate through and toss the array at the end?

Reply:
I guess not, then it would just get longer and longer.